executor/plan_replayer.go

```diff
@@ -61,101 +62,75 @@ const (
 // PlanReplayerExec represents a plan replayer executor.
 type PlanReplayerExec struct {
 	baseExecutor
-	ExecStmts []ast.StmtNode
-	Analyze   bool
-
-	endFlag bool
-}
-
-type tableNamePair struct {
-	DBName    string
-	TableName string
-	IsView    bool
+	DumpInfo *PlanReplayerDumpInfo
+	endFlag  bool
 }
 
-type tableNameExtractor struct {
-	ctx      context.Context
-	executor sqlexec.RestrictedSQLExecutor
-	is       infoschema.InfoSchema
-	curDB    model.CIStr
-	names    map[tableNamePair]struct{}
-	cteNames map[string]struct{}
-	err      error
+type PlanReplayerDumpInfo struct {
+	ExecStmts []ast.StmtNode
+	Analyze   bool
+	Path      string
+	File      *os.File
+	FileName  string
+	ctx       sessionctx.Context
 }
 
-func (tne *tableNameExtractor) Enter(in ast.Node) (ast.Node, bool) {
-	if _, ok := in.(*ast.TableName); ok {
-		return in, true
+// Next implements the Executor Next interface.
+func (e *PlanReplayerExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.GrowAndReset(e.maxChunkSize)
+	if e.endFlag {
+		return nil
 	}
-	return in, false
-}
-
-func (tne *tableNameExtractor) Leave(in ast.Node) (ast.Node, bool) {
-	if tne.err != nil {
-		return in, true
+	err := e.createFile(domain.GetPlanReplayerDirName())
+	if err != nil {
+		return err
 	}
-	if t, ok := in.(*ast.TableName); ok {
-		isView, err := tne.handleIsView(t)
+	if len(e.DumpInfo.Path) > 0 {
+		err = e.prepare()
 		if err != nil {
-			tne.err = err
-			return in, true
-		}
-		tp := tableNamePair{DBName: t.Schema.L, TableName: t.Name.L, IsView: isView}
-		if tp.DBName == "" {
-			tp.DBName = tne.curDB.L
-		}
-		if _, ok := tne.names[tp]; !ok {
-			tne.names[tp] = struct{}{}
-		}
-	} else if s, ok := in.(*ast.SelectStmt); ok {
-		if s.With != nil && len(s.With.CTEs) > 0 {
-			for _, cte := range s.With.CTEs {
-				tne.cteNames[cte.Name.L] = struct{}{}
-			}
+			return err
 		}
+		e.endFlag = true
+		e.ctx.GetSessionVars().LastPlanReplayerToken = e.DumpInfo.FileName
+		return nil
```

Comment:
As we can only read file from `handleSpecialQuery`, thus we store the file token in the seesion var return return nil here.

Reply:
Maybe we can add the comment in code.